File: src/main/java/com/mycompany/pt/Parser.java
/*
 * |-------------------------------------------------
 * | Copyright © 2017 Colin But. All rights reserved.
 * |-------------------------------------------------
 */
package com.mycompany.pt;

import com.mycompany.pt.model.CurrencyCode;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;

@NoArgsConstructor(access = AccessLevel.PRIVATE)
public final class Parser {

    private static PaymentListing paymentListing = PaymentListing.getInstance();

    private static final Parser INSTANCE = new Parser();

    public static Parser getInstance() {
        return INSTANCE;
    }

    public void invoke(String input) {
        String[] lineItems = input.split(" ");
        CurrencyCode currencyCode = CurrencyCode.valueOf(lineItems[0]);
        BigDecimal amount = new BigDecimal(lineItems[1]);
        paymentListing.addPayment(currencyCode, amount);
    }
}

File: kotlin/src/main/kotlin/payment/tracker/Parser.kt
/*
 * |-------------------------------------------------
 * | Copyright © 2019 Colin But. All rights reserved.
 * |-------------------------------------------------
 */
package payment.tracker

import payment.tracker.model.CurrencyCode
import java.lang.RuntimeException
import java.math.BigDecimal

class Parser {

    private val paymentListing : PaymentListing = PaymentListing(HashMap())

    fun invoke(input : String) {
        val lineItems = input.split(" ")
        if (lineItems.size != 2) {
            throw RuntimeException("Invalid arguments entered, expecting format [CurrencyCode Amount]")
        }

        if (lineItems[0] == "" || lineItems[1] == "") {
            throw RuntimeException("Invalid arguments entered, expecting format [CurrencyCode Amount]")
        }

        val currencyCode = CurrencyCode.valueOf(lineItems[0])
        val amount = BigDecimal(lineItems[1])

        paymentListing.addPayment(currencyCode, amount)
    }
}
